File: py/hubspotsettings_sdk/features.py
# HubspotSettings SDK feature factory

from hubspotsettings_sdk.feature.base_feature import HubspotSettingsBaseFeature
from hubspotsettings_sdk.feature.debug_feature import HubspotSettingsDebugFeature
from hubspotsettings_sdk.feature.idempotency_feature import HubspotSettingsIdempotencyFeature
from hubspotsettings_sdk.feature.metrics_feature import HubspotSettingsMetricsFeature
from hubspotsettings_sdk.feature.paging_feature import HubspotSettingsPagingFeature
from hubspotsettings_sdk.feature.ratelimit_feature import HubspotSettingsRatelimitFeature
from hubspotsettings_sdk.feature.retry_feature import HubspotSettingsRetryFeature
from hubspotsettings_sdk.feature.test_feature import HubspotSettingsTestFeature
from hubspotsettings_sdk.feature.timeout_feature import HubspotSettingsTimeoutFeature


_FEATURES = {
    "base": lambda: HubspotSettingsBaseFeature(),
    "debug": lambda: HubspotSettingsDebugFeature(),
    "idempotency": lambda: HubspotSettingsIdempotencyFeature(),
    "metrics": lambda: HubspotSettingsMetricsFeature(),
    "paging": lambda: HubspotSettingsPagingFeature(),
    "ratelimit": lambda: HubspotSettingsRatelimitFeature(),
    "retry": lambda: HubspotSettingsRetryFeature(),
    "test": lambda: HubspotSettingsTestFeature(),
    "timeout": lambda: HubspotSettingsTimeoutFeature(),
}


def _make_feature(name):
    factory = _FEATURES.get(name)
    if factory is not None:
        return factory()
    return _FEATURES["base"]()


# True when this SDK was generated with the named feature class - the
# constructor's tolerance for extend-carried features reads this (an
# active name with no generated class must not become a BaseFeature
# stray when an extend instance carries it).
def _has_feature(name):
    return name in _FEATURES

File: php/features.php
<?php
declare(strict_types=1);

// HubspotSettings SDK feature factory

require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/feature/DebugFeature.php';
require_once __DIR__ . '/feature/IdempotencyFeature.php';
require_once __DIR__ . '/feature/MetricsFeature.php';
require_once __DIR__ . '/feature/PagingFeature.php';
require_once __DIR__ . '/feature/RatelimitFeature.php';
require_once __DIR__ . '/feature/RetryFeature.php';
require_once __DIR__ . '/feature/TestFeature.php';
require_once __DIR__ . '/feature/TimeoutFeature.php';


class HubspotSettingsFeatures
{
    public static function make_feature(string $name)
    {
        switch ($name) {
            case "base":
                return new HubspotSettingsBaseFeature();
            case "debug":
                return new HubspotSettingsDebugFeature();
            case "idempotency":
                return new HubspotSettingsIdempotencyFeature();
            case "metrics":
                return new HubspotSettingsMetricsFeature();
            case "paging":
                return new HubspotSettingsPagingFeature();
            case "ratelimit":
                return new HubspotSettingsRatelimitFeature();
            case "retry":
                return new HubspotSettingsRetryFeature();
            case "test":
                return new HubspotSettingsTestFeature();
            case "timeout":
                return new HubspotSettingsTimeoutFeature();
            default:
                return new HubspotSettingsBaseFeature();
        }
    }

    /**
     * Does a generated feature class back this name? False for a name only
     * an options extend instance can supply (the station adopt path) - the
     * constructor uses this to skip make_feature for such names instead of
     * adding a stray BaseFeature.
     */
    public static function has_feature(string $name): bool
    {
        switch ($name) {
            case "base":
            case "debug":
            case "idempotency":
            case "metrics":
            case "paging":
            case "ratelimit":
            case "retry":
            case "test":
            case "timeout":
                return true;
            default:
                return false;
        }
    }
}
